SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKDevAuthTokenGenerator.cpp

```diff
@@ -0,0 +1,100 @@
+	: bIsGenerating(false)
+{
+}
+
+void FSpatialGDKDevAuthTokenGenerator::UpdateSettings(FString DevAuthToken)
+{
+	USpatialGDKEditorSettings* GDKEditorSettings = GetMutableDefault<USpatialGDKEditorSettings>();
+	GDKEditorSettings->DevelopmentAuthenticationToken = DevAuthToken;
+	GDKEditorSettings->SaveConfig();
+	GDKEditorSettings->SetRuntimeDevelopmentAuthenticationToken();
+	GDKEditorSettings->SetRuntimeUseDevelopmentAuthenticationFlow();
+}
+
+void FSpatialGDKDevAuthTokenGenerator::DoGenerateDevAuthTokenTasks()
+{
+	bool bIsRunningInChina = GetDefault<USpatialGDKSettings>()->IsRunningInChina();
+	AsyncTask(ENamedThreads::AnyBackgroundThreadNormalTask, [this, bIsRunningInChina]
+	{
+		AsyncTask(ENamedThreads::GameThread, [this]()
+		{
+			this->ShowTaskStartedNotification(TEXT("Generating Development Authentication Token"));
+		});
+
+		FString DevAuthToken;
+		if (SpatialCommandUtils::GenerateDevAuthToken(bIsRunningInChina, DevAuthToken))
+		{
+			AsyncTask(ENamedThreads::GameThread, [this, DevAuthToken]()
+			{
+				UpdateSettings(DevAuthToken);
+				this->ShowTaskEndedNotification(TEXT("Development Authentication Token Updated"), SNotificationItem::CS_Success);
+			});
+		}
+		else
+		{
+			UE_LOG(LogSpatialGDKDevAuthTokenGenerator, Error, TEXT("Failed to generate a development authentication token."));
+			AsyncTask(ENamedThreads::GameThread, [this]()
+			{
+				this->ShowTaskEndedNotification(TEXT("Failed to generate Development Authentication Token"), SNotificationItem::CS_Fail);
+			});
+		}
+	});
+}
+
+void FSpatialGDKDevAuthTokenGenerator::AsyncGenerateDevAuthToken()
+{
+	bool bExpected = false;
+	if (bIsGenerating.CompareExchange(bExpected, true))
+	{
+		DoGenerateDevAuthTokenTasks();
+	}
+	else
+	{
+		UE_LOG(LogSpatialGDKDevAuthTokenGenerator, Display, TEXT("A previous Development Authentication Token request is still pending. New request for generation ignored."));
+	}
+}
+
+void FSpatialGDKDevAuthTokenGenerator::ShowTaskStartedNotification(const FString& NotificationText)
+{
+	FNotificationInfo Info(FText::AsCultureInvariant(NotificationText));
+	Info.ExpireDuration = 5.0f;
+	Info.bFireAndForget = false;
+
+	TaskNotificationPtr = FSlateNotificationManager::Get().AddNotification(Info);
+
+	if (TaskNotificationPtr.IsValid())
+	{
+		TaskNotificationPtr.Pin()->SetCompletionState(SNotificationItem::CS_Pending);
+	}
+}
+
+void FSpatialGDKDevAuthTokenGenerator::ShowTaskEndedNotification(const FString& NotificationText, SNotificationItem::ECompletionState CompletionState)
+{
+	TSharedPtr<SNotificationItem> Notification = TaskNotificationPtr.Pin();
+	if (Notification.IsValid())
+	{
+		Notification->SetFadeInDuration(0.1f);
+		Notification->SetFadeOutDuration(0.5f);
+		Notification->SetExpireDuration(5.0);
+		Notification->SetText(FText::AsCultureInvariant(NotificationText));
+		Notification->SetCompletionState(CompletionState);
+		Notification->ExpireAndFadeout();
+	}
+	bIsGenerating = false;
```

Comment:
Looks like this is always being checked and changed from the game thread, so we don't actually need it to be atomic?

Reply:
So of course it wasn't atomic in the first iteration.. but given that this is a helper class, I may be overreacting, but the potential that someone calls it from a non-game thread means that it would be potential problem. As an atomic this will always be safe, not atomic this only works assuming the calling context is always the game thread. In this case I'd say the performance impact of using an atomic is insignificant and having the code "correct" in terms of enforcing that the single-instance behavior can not be compromised feels "right". It's a bit of defensive coding, but I believe for this case it's reasonable.